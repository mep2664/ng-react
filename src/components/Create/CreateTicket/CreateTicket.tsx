import * as React from "react";
import { Loader, SelectInput, TextArea } from "../..";
import styled from "styled-components";
import gql from "graphql-tag";
import { useMutation, useQuery } from "@apollo/react-hooks";

export const GET_PROJECTS = gql`
query {
    allProjects {
        projectName
    }
    allSprints {
        sprintName
    }
}
`;

export const CREATE_TICKET = gql`
    mutation CreateTicket($description: String, $priority: String,
        $sprintName: String, $projectName: String, $storyPoints: Int, $ticketType: String) {
            createTicket(projectName: $projectName, description: $description, priority: $priority,
                sprintName: $sprintName, storyPoints: $storyPoints, ticketType: $ticketType)
            {
                ticket {
                    ticketId
                    projectName
                    ticketNumber
                    description
                    priority
                    sprintName
                    storyPoints
                    ticketType
                }
            }
    }
`;

const TicketInfo = styled.div`
    display: grid;
    grid-gap: 25px;
    align-items: center;
    justify-items: end;
    grid-template-columns: minmax(auto, 300px) auto;
    box-sizing: border-box;
    padding: 25px;
`;

const SelectInputs = styled.div`
    display: grid;
    grid-gap: 10px;
    align-items: center;
    justify-items: end;
    grid-template-columns: auto minmax(auto, 300px);
`;

const DescriptionWrapper = styled.div`
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 100%;
`;

const typeOptions = [
    { caption: "Enhancement", value: "Enhancement" },
    { caption: "Bug", value: "Bug" },
    { caption: "Research", value: "Research" },
];

const priorityOptions = [
    { caption: "Blocker", value: "Blocker" },
    { caption: "Standard", value: "Standard" },
    { caption: "Minor", value: "Minor" },
];

const storyPointsOptions = () => {
    const sequence: number[] = [1, 2];
    for (let idx = 0; idx < 5; idx++) {
        sequence.push(sequence[idx] + sequence[idx + 1]);
    }
    const options: any[] = [];
    for (let numb in sequence) {
        options.push({ caption: sequence[numb].toString(), value: sequence[numb] });
    }
    return options;
}

export const CreateTicket: React.FC = () => {
    const [projectName, setProjectName] = React.useState<string>("");
    const [sprintName, setSprintName] = React.useState<string>("");
    const [ticketType, setTicketType] = React.useState<string>("");
    const [priority, setPriority] = React.useState<string>("");
    const [storyPoints, setStoryPoints] = React.useState<string>("");
    const [description, setDescription] = React.useState<string>("");
    const { loading, error, data } = useQuery(GET_PROJECTS);
    const [createTicket] = useMutation(CREATE_TICKET);

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const ticketData = {
            projectName,
            sprintName,
            ticketType,
            priority,
            storyPoints,
            description,
        }
        createTicket({ variables: ticketData })
            .catch((error) => alert(error.message));
    }

    if (loading) {
        return <Loader />;
    }

    if (error) {
        if (error.networkError && "statusCode" in error.networkError
            && error.networkError["statusCode"] === 405) {
            window.location.assign(`${window.location.protocol}//${window.location.host}/`);
        }
        return <div>{error.message}</div>;
    }

    const projectOptions = data.allProjects.map((project: any) => { return { caption: project.projectName, value: project.projectName } });
    const sprintOptions = data.allSprints.map((sprint: any) => { return { caption: sprint.sprintName, value: sprint.sprintName } });

    return (
        <div>
            {`Create Ticket...`}
            <TicketInfo>
                <form action="localhost:5556/ticket" method="post" onSubmit={handleSubmit}>
                    <SelectInputs>
                        <SelectInput label="Project" name="project" options={projectOptions} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setProjectName(e.target.value)} />
                        <SelectInput label="Sprint" name="sprint" options={sprintOptions} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setSprintName(e.target.value)} />
                        <SelectInput label="Type" name="type" options={typeOptions} value={ticketType} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setTicketType(e.target.value)} />
                        <SelectInput label="Priority" name="priority" options={priorityOptions} value={priority} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setPriority(e.target.value)} />
                        <SelectInput label="Story Points" name="storyPoints" options={storyPointsOptions()} value={storyPoints} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setStoryPoints(e.target.value)} />
                    </SelectInputs>
                    <DescriptionWrapper>
                        <TextArea label="Description" name="description" value={description} onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => setDescription(e.target.value)} />
                    </DescriptionWrapper>
                    <input type="submit" value="submit" />
                </form>
            </TicketInfo>
        </div>
    );
}