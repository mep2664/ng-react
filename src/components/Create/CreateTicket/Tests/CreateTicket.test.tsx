import React from "react"
import { render, fireEvent, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { CreateTicket, CREATE_TICKET, GET_PROJECTS } from "../CreateTicket"
import { MockedProvider } from '@apollo/react-testing';

jest.mock("../../../Loaders/Loader", () => {
    return {
        Loader: () => <span>loading...</span>
    };
})

const projects = [
    {
        projectName: "PROJECT-1",
    },
    {
        projectName: "PROJECT-2",
    },
    {
        projectName: "PROJECT-3",
    },
];

const sprints = [
    {
        sprintName: "SPRINT-1",
    },
    {
        sprintName: "SPRINT-2",
    },
    {
        sprintName: "SPRINT-3",
    },
];

const vars = {
    projectName: "PROJECT-3", sprintName: "SPRINT-2", ticketType: "Enhancement",
    priority: "Standard", storyPoints: "3", description: "NEW_DESCRIPTION",
};
let numTimesSubmitted = 0;
const mocks = [
    {
        request: {
            query: GET_PROJECTS,
            variables: {},
        },
        result: {
            data: {
                allProjects: projects,
                allSprints: sprints,
            },
        },
    },
    {
        request: {
            query: CREATE_TICKET,
            variables: vars,
        },

        result: () => {
            numTimesSubmitted++;
            return {
                data: {
                    createTicket: {
                        ticket: {
                            ticketId: "TICKET_ID",
                            projectName: vars.projectName,
                            ticketNumber: "1",
                            sprintName: vars.sprintName,
                            ticketType: vars.ticketType,
                            priority: vars.priority,
                            storyPoints: vars.storyPoints,
                            description: vars.description,
                        },
                    }
                },
            };
        },
    },
];

describe("Tests for the CreateTicket component", () => {
    test("Component matches snapshot", async () => {
        const { container } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        expect(container).toMatchSnapshot();
    });

    test("Displays loader when waiting for data", () => {
        const { getByText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        const loader = getByText("loading...");
        expect(loader).toBeDefined();
    });

    test("Displays error when request for data fails", async () => {
        const errorMock = {
            request: {
                query: GET_PROJECTS,
                variables: {},
            },
            error: new Error("something went wrong..."),
        };
        const { getByText } = render(
            <MockedProvider mocks={[errorMock]} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const error = getByText("Network error: something went wrong...");
        expect(error).toBeDefined();
    });

    test("Changes project name correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Project");
        fireEvent.change(input, { target: { value: vars.projectName } });
        expect(input).toHaveValue(vars.projectName);
    });

    test("Changes sprint name correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Sprint");
        fireEvent.change(input, { target: { value: vars.sprintName } });
        expect(input).toHaveValue(vars.sprintName);
    });

    test("Changes type correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Type");
        fireEvent.change(input, { target: { value: vars.ticketType } });
        expect(input).toHaveValue(vars.ticketType);
    });

    test("Changes priority correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Priority");
        fireEvent.change(input, { target: { value: vars.priority } });
        expect(input).toHaveValue(vars.priority);
    });

    test("Changes story points correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Story Points");
        fireEvent.change(input, { target: { value: vars.storyPoints } });
        expect(input).toHaveValue(vars.storyPoints);
    });

    test("Changes description correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Description");
        fireEvent.change(input, { target: { value: vars.description } });
        expect(input).toHaveValue(vars.description);
    });

    test("Submits correctly", async () => {
        const { getByText, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateTicket />
            </MockedProvider>,
        );
        await wait();

        const project = getByLabelText("Project");
        fireEvent.change(project, { target: { value: vars.projectName } });

        const sprint = getByLabelText("Sprint");
        fireEvent.change(sprint, { target: { value: vars.sprintName } });

        const type = getByLabelText("Type");
        fireEvent.change(type, { target: { value: vars.ticketType } });

        const priority = getByLabelText("Priority");
        fireEvent.change(priority, { target: { value: vars.priority } });

        const storyPoints = getByLabelText("Story Points");
        fireEvent.change(storyPoints, { target: { value: vars.storyPoints } });

        const description = getByLabelText("Description");
        fireEvent.change(description, { target: { value: vars.description } });

        const submit = getByText("submit");
        fireEvent.click(submit);
        await wait();
        expect(numTimesSubmitted).toEqual(1);
    });
});
