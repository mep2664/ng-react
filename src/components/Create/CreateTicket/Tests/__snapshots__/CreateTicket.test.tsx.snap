// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CreateTicket component Component matches snapshot 1`] = `
<div>
  <div>
    Create Ticket...
    <div
      class="sc-Rmtcm AwuUM"
    >
      <form
        action="localhost:5556/ticket"
        method="post"
      >
        <div
          class="sc-bRBYWo dCMlAG"
        >
          <label
            class="sc-ckVGcZ envGLI"
            for="selectInput__Project"
          >
            Project
          </label>
          <select
            class="sc-jKJlTe SQmOY"
            id="selectInput__Project"
            name="project"
          >
            <option
              value=""
            >
              &lt; Not Selected &gt;
            </option>
            <option
              value="PROJECT-1"
            >
              PROJECT-1
            </option>
            <option
              value="PROJECT-2"
            >
              PROJECT-2
            </option>
            <option
              value="PROJECT-3"
            >
              PROJECT-3
            </option>
          </select>
          <label
            class="sc-ckVGcZ envGLI"
            for="selectInput__Sprint"
          >
            Sprint
          </label>
          <select
            class="sc-jKJlTe SQmOY"
            id="selectInput__Sprint"
            name="sprint"
          >
            <option
              value=""
            >
              &lt; Not Selected &gt;
            </option>
            <option
              value="SPRINT-1"
            >
              SPRINT-1
            </option>
            <option
              value="SPRINT-2"
            >
              SPRINT-2
            </option>
            <option
              value="SPRINT-3"
            >
              SPRINT-3
            </option>
          </select>
          <label
            class="sc-ckVGcZ envGLI"
            for="selectInput__Type"
          >
            Type
          </label>
          <select
            class="sc-jKJlTe SQmOY"
            id="selectInput__Type"
            name="type"
          >
            <option
              value=""
            >
              &lt; Not Selected &gt;
            </option>
            <option
              value="Enhancement"
            >
              Enhancement
            </option>
            <option
              value="Bug"
            >
              Bug
            </option>
            <option
              value="Research"
            >
              Research
            </option>
          </select>
          <label
            class="sc-ckVGcZ envGLI"
            for="selectInput__Priority"
          >
            Priority
          </label>
          <select
            class="sc-jKJlTe SQmOY"
            id="selectInput__Priority"
            name="priority"
          >
            <option
              value=""
            >
              &lt; Not Selected &gt;
            </option>
            <option
              value="Blocker"
            >
              Blocker
            </option>
            <option
              value="Standard"
            >
              Standard
            </option>
            <option
              value="Minor"
            >
              Minor
            </option>
          </select>
          <label
            class="sc-ckVGcZ envGLI"
            for="selectInput__Story Points"
          >
            Story Points
          </label>
          <select
            class="sc-jKJlTe SQmOY"
            id="selectInput__Story Points"
            name="storyPoints"
          >
            <option
              value=""
            >
              &lt; Not Selected &gt;
            </option>
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="8"
            >
              8
            </option>
            <option
              value="13"
            >
              13
            </option>
            <option
              value="21"
            >
              21
            </option>
          </select>
        </div>
        <div
          class="sc-hzDkRC frjYVh"
        >
          <label
            class="sc-eNQAEJ hALvCQ"
            for="textArea__Description"
          >
            Description
          </label>
          <textarea
            class="sc-hMqMXs elAqlG"
            id="textArea__Description"
            name="description"
          />
        </div>
        <input
          type="submit"
          value="submit"
        />
      </form>
    </div>
  </div>
</div>
`;
